components/modules/matrix.c

```diff
@@ -0,0 +1,528 @@
+#define GET_PREV_STATUS(d) (d->queue[(d->write_offset - 2) & (QUEUE_SIZE - 1)])
+#define HAS_QUEUED_DATA(d) (d->read_offset < d->write_offset)
+#define HAS_QUEUE_SPACE(d) (d->read_offset + QUEUE_SIZE - 1 > d->write_offset)
+
+#define REPLACE_IT(d, x)                            \
+  (d->queue[(d->write_offset - 1) & (QUEUE_SIZE - 1)] = \
+       (matrix_event_t){(x), esp_timer_get_time()})
+#define QUEUE_IT(d, x)                            \
+  (d->queue[(d->write_offset++) & (QUEUE_SIZE - 1)] = \
+       (matrix_event_t){(x), esp_timer_get_time()})
+
+#define GET_READ_STATUS(d) (d->queue[d->read_offset & (QUEUE_SIZE - 1)])
+#define ADVANCE_IF_POSSIBLE(d)            \
+  if (d->read_offset < d->write_offset) { \
+    d->read_offset++;                     \
+  }
+
+static esp_err_t set_gpio_mode_input(int pin, gpio_int_type_t intr) {
+  gpio_config_t config = {.pin_bit_mask = 1LL << pin,
+                          .mode = GPIO_MODE_INPUT,
+                          .pull_up_en = GPIO_PULLUP_ENABLE,
+                          .pull_down_en = GPIO_PULLDOWN_DISABLE,
+                          .intr_type = intr};
+
+  return gpio_config(&config);
+}
+
+static esp_err_t set_gpio_mode_output(int pin) {
+  gpio_config_t config = {.pin_bit_mask = 1LL << pin,
+                          .mode = GPIO_MODE_OUTPUT_OD,
+                          .pull_up_en = GPIO_PULLUP_DISABLE,
+                          .pull_down_en = GPIO_PULLDOWN_DISABLE
+                        };
+
+  return gpio_config(&config);
+}
+
+static void set_columns(DATA *d, int level) {
+  for (int i = 0; i < d->column_count; i++) {
+    gpio_set_level(d->columns[i], level);
+  }
+}
+
+static void initialize_pins(lua_State *L, DATA *d) {
+  for (int i = 0; i < d->column_count; i++) {
+    if (set_gpio_mode_output(d->columns[i]) != ESP_OK) {
+      luaL_error(L, "Unable to configure pins");
+    }
+  }
+
+  set_columns(d, 0);
+
+  for (int i = 0; i < d->row_count; i++) {
+    if (set_gpio_mode_input(d->rows[i], GPIO_INTR_NEGEDGE) != ESP_OK) {
+      luaL_error(L, "Unable to configure pins");
+    }
+  }
+}
+
+static void disable_row_interrupts(DATA *d) {
+  for (int i = 0; i < d->row_count; i++) {
+    gpio_set_intr_type(d->rows[i], GPIO_INTR_DISABLE);
+  }
+}
+
+// Just takes the channel number. Cleans up the resources used.
+static int matrix_close(DATA *d) {
+  if (!d) {
+    return 0;
+  }
+
+  disable_row_interrupts(d);
+
+  for (int i = 0; i < d->row_count; i++) {
+    gpio_isr_handler_remove(d->rows[i]);
+  }
+
+  for (int i = 0; i < d->column_count; i++) {
+    set_gpio_mode_input(d->columns[i], GPIO_INTR_DISABLE);
+  }
+
+  return 0;
+}
+
+// Character returned is 0 .. max if pressed. -1 if not.
+static int matrix_get_character(DATA *d)
+{
+  set_columns(d, 1);
+  disable_row_interrupts(d);
+
+  int character = -1;
+
+  // We are either waiting for a negative edge (keypress) or a positive edge
+  // (keyrelease)
```

Comment:
This comment is confusing me - we just disabled row interrupts, so why would we be waiting for an edge?